akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/OffsetStoreDao.scala

```diff
@@ -113,6 +115,26 @@ import software.amazon.awssdk.services.dynamodb.model.WriteRequest
       }(ExecutionContext.parasitic)
   }
 
+  def writeWholeBatch(
+      batchReq: BatchWriteItemRequest,
+      maxRetriesOfUnprocessed: Int = 3): Future[List[BatchWriteItemResponse]] = {
+    val result = client.batchWriteItem(batchReq).asScala
+
+    result.flatMap { response =>
+      if (response.hasUnprocessedItems && maxRetriesOfUnprocessed > 0) {
+        val unprocessed = response.unprocessedItems
+        val newReq = batchReq.toBuilder.requestItems(unprocessed).build()
+        writeWholeBatch(newReq, maxRetriesOfUnprocessed - 1).map { responses =>
+          response :: responses
+        }(ExecutionContext.parasitic)
+      } else {
+        // TODO: return a partial failure?  for now, to check that everything was written, would have to inspect the
+        // last response for unprocessed items
+        Future.successful(List(response))
```

Comment:
Would be nicer to not need to repeat those checks everywhere.

Reply:
Yeah a custom exception is probably the most idiomatic thing here